docs/key_modules/memory.md

```diff
@@ -0,0 +1,315 @@
+# Memory
+
+## Overview
+
+The Memory module in CAMEL provides a flexible and powerful system for storing, retrieving, and managing information for AI agents. It enables agents to maintain context across conversations and retrieve relevant information from past interactions, enhancing the coherence and relevance of AI responses.
+
+## Getting Started
+
+### Installation
+
+Ensure you have CAMEL AI installed in your Python environment:
+
+```bash
+pip install camel-ai
+
+```
+
+### Basic Usage
+
+Here's a quick example of how to use the LongtermAgentMemory:
+
+```python
+from camel.memories import LongtermAgentMemory, ChatHistoryBlock, VectorDBBlock
+from camel.memories import MemoryRecord, ScoreBasedContextCreator
+from camel.messages import BaseMessage
+from camel.types import OpenAIBackendRole
+from camel.utils import OpenAITokenCounter
+
+
+# Initialize memory
+memory = LongtermAgentMemory(
+    context_creator=ScoreBasedContextCreator(OpenAITokenCounter, 1024),
+    chat_history_block=ChatHistoryBlock(),
+    vector_db_block=VectorDBBlock()
+)
+
+# Write records to memory
+records = [
+    MemoryRecord(
+        message=BaseMessage(role_name="user", role_type="human", meta_dict=None, content="What is AI?"),
+        role_at_backend=OpenAIBackendRole.USER
+    ),
+    MemoryRecord(
+        message=BaseMessage( role_name="agent", role_type="ai", meta_dict=None, content="AI refers to systems that mimic human intelligence."),
+        role_at_backend=OpenAIBackendRole.ASSISTANT
+    )
+]
+memory.write_records(records)
+
+# Retrieve context
+context, token_count = memory.get_context()
+print(f"Retrieved context (token count: {token_count}):")
+for message in context:
+    print(f"{message}")
+
+```
+
+## Core Components
+
+### MemoryRecord
+
+The basic unit of storage in the CAMEL memory system.
+
+Attributes:
+
+- `message`: The main content of the record (BaseMessage)
+- `role_at_backend`: The role this message played at the OpenAI backend (OpenAIBackendRole)
+- `uuid`: A unique identifier for the record
+- `extra_info`: Additional key-value pairs for extra information
+
+Methods:
+
+- `from_dict()`: Reconstruct a MemoryRecord from a dictionary
+- `to_dict()`: Convert the MemoryRecord to a dictionary for serialization
+- `to_openai_message()`: Convert the record to an OpenAIMessage object
+
+### ContextRecord
+
+Represents the result of memory retrieval.
+
+Attributes:
+
+- `memory_record`: A MemoryRecord
+- `score`: A float value representing the relevance or importance of the record
+
+### MemoryBlock (Abstract Base Class)
+
+Serves as the fundamental component within the agent memory system.
```

Comment:
```suggestion
Serves as the fundamental component within the agent memory system. `MemoryBlock` class follows the "Composite Design pattern" so that you can compose objects into tree structures and then work with these structures as if they were individual objects.
```